src/Nethermind/Nethermind.Blockchain/OnChainTxWatcher.cs

```diff
@@ -58,22 +59,45 @@ private void OnBlockAddedToMain(object sender, BlockReplacementEventArgs e)
         private void ProcessBlock(Block block, Block? previousBlock)
         {
             _txPool.BlockGasLimit = block.GasLimit;
+            _txPool.CurrentBaseFee = block.Header.BaseFee;
             long transactionsInBlock = block.Transactions.Length;
             long discoveredForPendingTxs = 0;
             long discoveredForHashCache = 0;
+            HashSet<Address> senders = new();
             
             for (int i = 0; i < transactionsInBlock; i++)
             {
-                Keccak txHash = block.Transactions[i].Hash;
+                Transaction tx = block.Transactions[i];
+                Keccak txHash = tx.Hash;
+                
                 if (!_txPool.IsInHashCache(txHash))
                 {
                     discoveredForHashCache++;
                 }
-                if (!_txPool.RemoveTransaction(txHash, true))
+
+                if (!_txPool.RemoveTransaction(tx, true))
                 {
                     discoveredForPendingTxs++;
                 }
+                
+                senders.Add(tx.SenderAddress);
             }
+
+            Transaction[]? pendingTxs = _txPool.GetPendingTransactions();
+
+            foreach (Transaction tx in pendingTxs)
+            {
+                if (tx?.Type == TxType.EIP1559)
+                {
+                    senders.Add(tx.SenderAddress);
```

Comment:
I don't understand this change. Could explain that?

Reply:
Sure. After every processed block we want to recalculate buckets of addresses:
1) with removed transactions - we need to check if new first-to-execute tx has sufficient balance and to calculate GasBottleneck
2) with 1559 txs - we need to recalculate GasBottleneck after every block because of baseFee changes

As one address can have more than one tx in block, we don't want to recalculate bucket after every removal - we are adding addresses of changed buckets to HashSet "senders". We are adding every address having at least one 1559 tx in bucket to this HashSet too. At the end we do RemoveOrUpdateBucket for every address in senders